Introduce line breaking in format command

sorenhartmann opened the pull request:
WIP

sorenhartmann committed:
Initial line wrapping implementation

Commit:
Some queries for dealing with &'s and indentation

Commit:
Make clippy happy

ZedThree reviewed (changes requested):
Thanks @sorenhartmann! This looks great, it's really appreciated!

I have some high level comments:

- it would be good to get some unit tests on some of the individual functions -- I'm happy to lend a hand with this
- after fixing writing the output, I notice there are some test failures: it looks like a line break is being forced in cases where one isn't necessary? For example, this snippet:
```diff
-  function real_not(a)
+  function real_not( &
+    a)
```
Ahh, this is because the line length is currently hardcoded to `20`, with an indent of `2`; changing this to `80` and all the tests pass -- ok, so this is just a matter of needing to handle user options. Again, something I'm happy to help with
- we'll have to handle the different newline styles -- `\n`/`\r\n`/`\r`, but we can put this off for now
- 
- this "black-esque" style, with the vertical wrapping, is not going to be to everyone's tastes -- we've had a fair bit of discussion about this. I do like it, because it basically requires no options, and so there can be no arguments about tuning it really. However, we'll likely end up having this as one of two options, with the other being a flatter wrapping style -- just something to bear in mind

Also, a couple of points to help with developing rust:
- it's good to run `cargo clippy` fairly frequently (or even better, set up LSP or something in your editor to run it automatically). This is the rust linter, and it's _fantastic_, it's really helpful for learning rust.
- `README.dev.md` describes how to run the tests

Commit:
Properly apply iterative method instead of applying only twice

Commit:
Use logging macros

Commit:
Update some wrapping queries

Commit:
Update some topiary queries

sorenhartmann committed:
Rename method

sorenhartmann committed:
Some more line splitting examples

Courtesy of ChatGPT

ZedThree committed:
Write formatted source to `output`

ZedThree committed:
Add long-lines test case

ZedThree committed:
Rename `FormatArgs` to `FormatCommand`

ZedThree committed:
Add structs for format settings

These fit into the `Configuration` framework. Some work still needs to
be done, namely that `Settings` is currently living in the linter
crate, when it should be in the workspace crate.

sorenhartmann committed:
Remove unused file left-over from rebase

sorenhartmann commented on fortitude/resources/test/fixtures/format/long-lines.f90:
Yes, that is correct

sorenhartmann commented:
@ZedThree @LiamPattinson Thank you for the feedback, it is really appreciated!

First of all, I apologize for the late reply.
 
- I am not really used to the GitHub workflow, but I have tried to respond to your comments above. Also, for rebasing, should I just force-push this branch?
- I have updated the topiary queries to handle some more constructs, and also handle some indentation across line-breaks.
- As far as I understand, according to https://github.com/PlasmaFAIR/fortitude/issues/3, `ruff` may also be able to handle formatting / line breaking to some extend? This seems very interesting! However, maybe we/I should wait with further development on this PR until the scope of a `ruff` solution is more clear?

ZedThree commented:
Thanks @sorenhartmann! Seems like the clippy issue is resolved, so no need to rebase.

We've got a session at a workshop next week, so I might see if we can get some people to try out this branch and get some real world feedback.

> As far as I understand, according to https://github.com/PlasmaFAIR/fortitude/issues/3, ruff may also be able to handle formatting / line breaking to some extend? This seems very interesting! However, maybe we/I should wait with further development on this PR until the scope of a ruff solution is more clear?

That's probably wise, I would really very much hate to waste your time! One thing about using ruff's solution would be that we can take advantage of any improvements they make very easily, but it's possible that it's a lot of work to use it in the first place.

I'm also not sure how flexible their solution is. If it's a bunch of work to achieve the same results you've got here, then it is probably not worth it.

ZedThree committed:
Apply cargo fmt
